docs(general): update contributing guide and docs (#1899)
openclarity/vmclarity: CONTRIBUTING.md

## Development

> After cloning / forking the repository, take a look at the `make` targets that are used for checking, generating,
> building and publishing code that can be listed by running `make help`.

Reviewer: ```suggestion
After cloning the repository, you can run `make help` to inspect the targets that are used for checking, generating,
building, and publishing code.
```

### Dependencies

- `Docker` (for local and e2e testing)
- `Go` for the backend (the current version used by the project can be found in the .go-version file)
- A Node package manager for the frontend, such as `npm` or `yarn`

> Most further dependencies of the `make` targets are automatically installed if not present locally.

Reviewer: ```suggestion
Internal dependencies by `make` targets are automatically installed if not present locally.
```

### Development Environment

Depending on your IDE/editor of choice, you might need a `go.work` file for the `gopls` language server to find all
references properly, such as:

```text
go 1.22.4

use (
  ./api/client
  ./api/server
  ./api/types
  ./cli
  ./containerruntimediscovery/client
  ./containerruntimediscovery/server
  ./containerruntimediscovery/types
  ./core
  ./e2e
  ./e2e/testdata
  ./installation
  ./orchestrator
  ./provider
  ./plugins/runner
  ./plugins/sdk-go
  ./plugins/sdk-go/example
  ./plugins/store/kics
  ./scanner
  ./testenv
  ./uibackend/client
  ./uibackend/server
  ./uibackend/types
  ./utils
    ./workflow)
```

> ### Running the VMClarity stack locally

Reviewer: ```suggestion
### Running the VMClarity stack locally using Docker
```

> For trying out the changes in the code, the VMClarity stack can be ran with the Docker provider locally, after the

Reviewer: ```suggestion
For testing the changes across the whole stack, VMClarity can be ran with Docker provider locally, after the
```

images have been [built](#building-vmclarity-containers) and their tags have been updated in the
`installation/docker/image_override.env` file:

```shell

> docker compose --project-name vmclarity --file installation/docker/docker-compose.yml --env-file
> installation/docker/image_override.env up -d --wait --remove-orphans

Reviewer: ```suggestion
docker compose --project-name vmclarity \
		--file installation/docker/docker-compose.yml \
		--env-file installation/docker/image_override.env \
		up -d --wait --remove-orphans
```

```

> When developing only one component of the stack, the component in question can be commented out in the [docker compose

Reviewer: ```suggestion
When working only on one stack component, the component in question can be commented out in the [docker compose
```

file](installation/docker/docker-compose.yml) and ran separately with `go run`, or in the case of the UI, with the
following commands:

```shell
cd ui
npm install
npm start
```

> Also pay attention to modify the [NGINX config](installation/docker/gateway.conf) accordingly!

Reviewer: ```suggestion
Update the changes in [NGINX config](installation/docker/gateway.conf) accordingly if they components in question are affected to ensure that Docker can communicate with them if they are ran on local network.
```

Author: What about this?

```markdown
Update the [NGINX config](installation/docker/gateway.conf) accordingly if the components in question are affected to
ensure that Docker can communicate with them if they are ran on local network.
Some environment variables could also be necessary for you to export in your shell before running the component, inspect
the contents of the corresponding `.env` file in the `installation/docker` directory!
```

To clean up the VMClarity stack locally, run:

```shell
docker compose --project-name vmclarity --file installation/docker/docker-compose.yml down --remove-orphans --volumes
```

### Building VMClarity Binaries

Makefile targets are provided to compile and build the VMClarity binaries. `make build` can be used to build all the

> components, `make build-all-go` only builds the go modules and `make ui` is for only building the UI.

Reviewer: ```suggestion
components, while `make build-all-go` and `make ui` only build the go modules or the UI.
```

### Building VMClarity Containers

`make docker` can be used to build the VMClarity containers for all the components. Specific targets for example `make
docker-cli` and `make docker-ui-backend` are also provided.

In order to also publish the VMClarity containers to a registry, please set the `DOCKER_PUSH` environment variable to
`true`. You can override the destination registry as well:

```shell

> DOCKER_REGISTRY=docker.io/tehsmash DOCKER_PUSH=true make docker

Reviewer: ```suggestion
DOCKER_REGISTRY=docker.io/my-vmclarity-images DOCKER_PUSH=true make docker
```

```

You must be logged into the docker registry locally before using this target.

### Linting

`make lint` can be used to run all the required linting rules over the code. In this case, the following targets will be
ran:

- `make license-check` can be used to validate that all the files in the repo have the correctly formatted license
header.
- `make lint-actions` checks Github Actions workflow files.
- `make lint-bicep` lints Bicep files.
- `make lint-cfn` lints Cloudformation files.
- `make lint-go` runs `golangci-lint` on the Go files. Rules and config can be viewed in the `.golangci.yml` file in the
root of the repo.
- `make lint-helm` lints the Helm chart.

> `make fix` is also provided which will resolve lint issues which are automatically fixable for example format issues.

Reviewer: ```suggestion
`make fix` is also provided which can automatically resolve lint issues such as formatting.
```

### Unit tests

`make test` can be used run all the unit tests in the repo. Alternatively you can use the standard go test CLI to run a
specific package or test like:

```shell
go test ./cli/cmd/... -run Test_isSupportedFS
```

### Generators

`make gen` runs the following targets that can be ran separately as well:

- After making changes to the API schema in `api/openapi.yaml`, you can run `make gen-api` to regenerate the model,
client and server code.
- Run `make gen-bicep` for generating bicep files after modifying them for installing VMClarity on Azure.
- Run `make gen-helm-docs` for generating the docs after making changes to VMClarity's Helm chart.

### Testing End to End

> `make e2e-docker` can be used run the end-to-end tests in the repository locally.

Reviewer: ```suggestion
`make e2e-docker` can be used run the end-to-end tests in the repository locally using Docker. `make e2e-k8s` can also be used to run end-to-end tests for Kubernetes provider using Docker.
```
